### PR TITLE
Oracle Provider

### PR DESCRIPTION
**Pull request by totycro**
# Overview
Oracle Provider for pygeoapi implemented by @xkosubek

The implementation is modeled after the PostgreSQL provider and thus should be fairly consistent with related code in pygeoapi. 

It supports all CRUD operations.

The tests are executed via GitHub Actions using an oracle docker container.

There is some documentation, specifically also for the `SqlManipulator`, which is a mechanism for custom extensions that is necessary for a certain private use case, but is designed to be useable for different contexts.

(as per https://github.com/geopython/pygeoapi/blob/master/CONTRIBUTING.md#contributions-and-licensing)

- [X] I'd like to contribute [feature X|bugfix Y|docs|something else] to pygeoapi. I confirm that my contributions to pygeoapi will be compatible with the pygeoapi license guidelines at the time of contribution.
- [X] I have already previously agreed to the pygeoapi Contributions and Licensing Guidelines

### CODE REVIEW

File: docs/source/data-publishing/ogcapi-features.rst [hunk 1]
```diff
@@ -258,6 +258,129 @@ MongoDB
...
+        # target_crs: 31287 # defaults to 4326 if not provided
+
+The provider supports connection over host and port with SID or SERVICE_NAME. For TNS naming, the system 
+environment variable TNS_ADMIN or the configuration parameter tns_admin must be set.
+
+The providers supports external authentication. At the moment only wallet authentication is implemented.
+
+Sometimes it is necessary to use the Oracle client for the connection. In this case init_oracle_client must be set to True.
+
+The provider supports a SQL-Manipulator-Plugin class. With this, the SQL statement could be manipulated. This is
```
Comment (tomkralidis) [inline docs/source/data-publishing/ogcapi-features.rst:305]: Perhaps this level of information (code blocks, etc.) can go in `pygeoapi/provider/oracle.py` and specify in the doc to consult the plugin itself?  We should keep these sections of the docs more user focused.

File: oracle-config.yml
Comment (tomkralidis) [inline oracle-config.yml:1]: Remove this file (given we have an example in `tests`.

File: oracle-openapi.yml
Comment (tomkralidis) [inline oracle-openapi.yml:1]: Remove this file.

File: pygeoapi/provider/oracle.py [hunk 2]
```diff
@@ -0,0 +1,1027 @@
...
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+# OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+# HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# =================================================================
+
```
Comment (tomkralidis) [inline pygeoapi/provider/oracle.py:29]: Sort dependencies as standard library, external packages, local imports, alphabetically.  So:

```python
import importlib
import json
import logging
from typing import Optional

import oracledb

from pygeoapi.provider.base import (
    BaseProvider,
    ProviderConnectionError,
    ProviderItemNotFoundError,
    ProviderQueryError
)
```

File: pygeoapi/provider/oracle.py [hunk 3]
```diff
@@ -0,0 +1,1027 @@
...
+        if properties:
+            prop_clauses = [f"{key} = :{key}" for key, value in properties]
+            where_conditions += prop_clauses
+            where_dict["properties"] = dict(properties)
+
+        if bbox:
+            bbox_dict = {"clause": "", "properties": {}}
+
+            sdo_mask = "mask={}".format(sdo_mask)
```
Comment (tomkralidis) [inline pygeoapi/provider/oracle.py:314]: use f-string

File: pygeoapi/provider/oracle.py [hunk 4]
```diff
@@ -0,0 +1,1027 @@
...
+                                                      :miny, \
+                                                      :maxx, \
+                                                      :maxy)), \
+                             :sdo_mask) = 'TRUE'"
+
+            where_conditions.append(bbox_dict["clause"])
+            where_dict["properties"].update(bbox_dict["properties"])
+
+        if where_conditions:
+            where_dict["clause"] = " WHERE {}".format(
```
Comment (tomkralidis) [inline pygeoapi/provider/oracle.py:346]: use f-string

File: pygeoapi/provider/oracle.py [hunk 5]
```diff
@@ -0,0 +1,1027 @@
...
+
+    :param module_class_string: full name of the class to create an object of
+    :param super_cls: expected super class for validity, None if bypass
+    :param kwargs: parameters to pass
+    :return instance of the given class:
+    """
+    module_name, class_name = module_class_string.rsplit(".", 1)
+    module = importlib.import_module(module_name)
+    LOGGER.debug(
+        "reading class {} from module {}".format(class_name, module_name)
```
Comment (tomkralidis) [inline pygeoapi/provider/oracle.py:1016]: use f-string

File: pygeoapi/provider/oracle.py [hunk 6]
```diff
@@ -0,0 +1,1027 @@
...
+    )
+    cls = getattr(module, class_name)
+    if super_cls is not None:
+        assert issubclass(
+            cls, super_cls
+        ), "class {} should inherit from {}".format(
```
Comment (tomkralidis) [inline pygeoapi/provider/oracle.py:1022]: use f-string

File: pygeoapi/provider/oracle.py [hunk 7]
```diff
@@ -0,0 +1,1027 @@
...
+            class_name, super_cls.__name__
+        )
+    LOGGER.debug("initialising {} with params {}".format(class_name, kwargs))
```
Comment (tomkralidis) [inline pygeoapi/provider/oracle.py:1025]: use f-string

**Review by tomkralidis — CHANGES_REQUESTED** [review]
Excellent work here and super easy/pleasant to review.  See minor change requests.  Thanks!

**Comment by xkosubek** [comment]
Thanks for the laud! I'll try to fix everything in the next days.

File: docs/source/data-publishing/ogcapi-features.rst (hunk 1 above)
Comment (tomkralidis) [reply docs/source/data-publishing/ogcapi-features.rst:305]: +1

File: oracle-config.yml
Comment (webb-ben) [reply oracle-config.yml:1]: If the full configuration files are beneficial to keep, alternatively consider creating an [oracle docker example](https://github.com/geopython/pygeoapi/tree/master/docker/examples).

File: tests/pygeoapi-test-config-oracle.yml
Comment (webb-ben) [inline tests/pygeoapi-test-config-oracle.yml:1]: @tomkralidis I believe this file can be deleted too - https://github.com/geopython/pygeoapi/pull/1329/files#diff-bec70aa2812a4deafc7d4f626335d69a1467963368c7b7946d62462d38abe7afR121

File: oracle-config.yml
Comment (xkosubek) [reply oracle-config.yml:1]: No, there is no benefit. Everything is in the documentation, too. I only forgot to delete this file...

File: tests/pygeoapi-test-config-oracle.yml
Comment (xkosubek) [reply tests/pygeoapi-test-config-oracle.yml:1]: Yes, the configuration is included in the test class.

**Comment by xkosubek** [comment]
All change requests are now implemented and commited!

**Pull request merged**